examples/http_auth.rs

```diff
@@ -0,0 +1,276 @@
+    let nonce = rng.gen();
+    let req_pq = schema::rpc::req_pq {
+        nonce: nonce,
+    };
+
+    let http_request = tryf!(create_http_request(&mut session, req_pq, MessageType::PlainText));
+    let auth_future = future_request(&http_client, http_request).and_then(move |response_bytes|
+        -> Box<Future<Item = (Vec<u8>, Session), Error = error::Error>>
+    {
+        let response: Message<schema::ResPQ> = tryf!(parse_response(&mut session, &response_bytes));
+
+        let res_pq = response.unwrap_plain_text_body();
+
+        if nonce != res_pq.nonce {
+            bailf!(ErrorKind::NonceMismatch(nonce, res_pq.nonce));
+        }
+
+        let pq_u64 = BigEndian::read_u64(&res_pq.pq);
+        println!("Decomposing pq = {}...", pq_u64);
+        let (p_u32, q_u32) = tryf!(asymm::decompose_pq(pq_u64));
+        println!("Decomposed p = {}, q = {}", p_u32, q_u32);
+        let u32_to_vec = |num| {
+            let mut v = vec![0; 4];
+            BigEndian::write_u32(v.as_mut_slice(), num);
+            v
+        };
+        let p = u32_to_vec(p_u32);
+        let q = u32_to_vec(q_u32);
+
+        let p_q_inner_data = schema::P_Q_inner_data {
+            pq: res_pq.pq,
+            p: p.clone().into(),
+            q: q.clone().into(),
+            nonce: res_pq.nonce,
+            server_nonce: res_pq.server_nonce,
+            new_nonce: rng.gen(),
+        };
+
+        println!("Data to send: {:#?}", &p_q_inner_data);
+        let p_q_inner_data_serialized = tryf!(serde_mtproto::to_bytes(&p_q_inner_data));
+        println!("Data bytes to send: {:?}", &p_q_inner_data_serialized);
+        let known_sha1_fingerprints = tryf!(asymm::KNOWN_RAW_KEYS.iter()
+            .map(|raw_key| {
+                let sha1_fingerprint = raw_key.read()?.sha1_fingerprint()?;
+                Ok(sha1_fingerprint.iter().map(|b| format!("{:02x}", b)).collect::<String>())
+            })
+            .collect::<error::Result<Vec<_>>>());
+        println!("Known public key SHA1 fingerprints: {:?}", known_sha1_fingerprints);
+        let known_fingerprints = tryf!(asymm::KNOWN_RAW_KEYS.iter()
+            .map(|raw_key| Ok(raw_key.read()?.fingerprint()?))
+            .collect::<error::Result<Vec<_>>>());
+        println!("Known public key fingerprints: {:?}", known_fingerprints);
+        let server_pk_fingerprints = res_pq.server_public_key_fingerprints.inner().as_slice();
+        println!("Server public key fingerprints: {:?}", &server_pk_fingerprints);
+        let (rsa_public_key, fingerprint) =
+            tryf!(asymm::find_first_key_fail_safe(server_pk_fingerprints));
+        println!("RSA public key used: {:#?}", &rsa_public_key);
+        let encrypted_data = tryf!(rsa_public_key.encrypt(&p_q_inner_data_serialized));
+        println!("Encrypted data: {:?}", encrypted_data.as_ref());
+        let encrypted_data2 = tryf!(rsa_public_key.encrypt2(&p_q_inner_data_serialized));
+        println!("Encrypted data 2: {:?}", &encrypted_data2);
```

Comment:
I'm assuming these are just for debugging purposes. Would you like to add the `log` crate and change these to `trace!` perhaps?

Reply:
Previously I thought "since this is an example, `println!` is ought to be a thing if you want to explain what is happening". But after both you and @habnabit pointed out, I guess it does make sense to use log calls: they can be switched on/off as you wish, `println!`s cannot.